theano/sandbox/mkl/__init__.py

```diff
@@ -0,0 +1,254 @@
+from __future__ import absolute_import, print_function, division
+import os
+import errno
+import sys
+import logging
+import textwrap
+import stat
+import shutil
+import theano
+from theano import config, gof
+from theano.gof.cmodule import Compiler, get_lib_extension, GCC_compiler
+from theano.sandbox.mkl.mkl_helper import header_text
+from theano.gof.compilelock import get_lock, release_lock
+
+from theano.gof import EquilibriumDB, SequenceDB
+
+from theano.tensor.blas import ldflags
+
+_logger_name = 'theano.sandbox.mkl'
+_logger = logging.getLogger(_logger_name)
+
+
+mkl_optimizer = EquilibriumDB(ignore_newtrees=False)
+mkl_seqopt = SequenceDB()
+
+
+def register_opt(*tags, **kwargs):
+    if any([not isinstance(t, str) for t in tags]):
+        raise RuntimeError("Bad call to register_opt."
+                           " All tags must be strings.", tags)
+
+    def f(local_opt):
+        name = (kwargs and kwargs.pop('name')) or local_opt.__name__
+        mkl_optimizer.register(name, local_opt, 'fast_run', 'fast_compile',
+                               'mkl', *tags, **kwargs)
+        return local_opt
+    return f
+
+
+class MKLVersion(gof.Op):
+    def c_header_dirs(self):
+        return ldflags(libs=False, include_dir=True)
+
+    def c_libraries(self):
+        return ldflags()
+
+    def c_lib_dirs(self):
+        return ldflags(libs=False, libs_dir=True)
+
+    def make_node(self):
+        return gof.Apply(self, [], [gof.Generic()()])
+
+    def c_support_code(self):
+        ccode = header_text()
+        ccode += """
+        #if PY_MAJOR_VERSION >= 3
+            #define PyInt_FromLong PyLong_FromLong
+        #endif
+        """
+        return ccode
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        o = outputs[0]
+        return textwrap.dedent(
+            """
+            MKLVersion v;
+            mkl_get_version(&v);
+            %(o)s = PyInt_FromLong(atoi(v.Build));
+            """) % locals()
+
+    def c_code_cache_version(self):
+        # Not needed, but make it clear that we do not want to cache this.
+        return None
+
+
+def mkl_version():
+    """
+    Return the current mkl version (e.g., 20160701) we compile with.
+    While `0` indicate fail to get mkl version.
+    """
+    if mkl_version.v is None:
+        try:
+            f = theano.function([], MKLVersion()(),
+                                theano.Mode(optimizer=None),
+                                profile=False)
+            mkl_version.v = f()
+        except:
```

Comment:
Why this try except? If there is an error here, is that a sign of a problem we want to raise to the user?

If this need to stay, you should always catch some exception, otherwise, it will also catch interruption like ctrl-C and we don't want that. Be the most precise in which exception you catch (Exception is better then not setting one. But ValueError is better then Exception).